src/Compilers/Core/Portable/SourceGeneration/Nodes/PredicateSyntaxStrategy.cs

```diff
@@ -84,14 +84,24 @@ public void SaveStateAndFree(StateTableStore.Builder tables)
                     {
                         var stopwatch = SharedStopwatch.StartNew();
                         var nodes = getFilteredNodes(root.Value, _owner._filterFunc, cancellationToken);
-                        entry = _filterTable.AddEntries(nodes, state, stopwatch.Elapsed, noInputStepsStepInfo, state);
+
+                        if (state != EntryState.Modified || !_filterTable.TryModifyEntries(nodes, Roslyn.Utilities.ReferenceEqualityComparer.Instance, stopwatch.Elapsed, noInputStepsStepInfo, state, out entry))
+                        {
+                            entry = _filterTable.AddEntries(nodes, state, stopwatch.Elapsed, noInputStepsStepInfo, state);
+                        }
                     }
 
                     // now, using the obtained syntax nodes, run the transform
-                    foreach (SyntaxNode node in entry)
+                    for (var i = 0; i < entry.Count; i++)
                     {
+                        if (entry.GetState(i) == EntryState.Removed)
```

Comment:
Why do we need to use `GetState` directly? I thought the `TableEntry` enumerator included removed entries?

Reply:
Yes, but the enumerator returns only the values, I need to know also the state here.